tensorflow/core/kernels/mkl/mkl_sparse_matrix_matmul_op.cc

```diff
@@ -0,0 +1,225 @@
+/* Copyright 2024 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
+
+#define EIGEN_USE_THREADS
+
+#include "Eigen/Core"
+#include "Eigen/SparseCore"
+#include "dnnl.hpp"
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/tensor_types.h"
+#include "tensorflow/core/framework/type_traits.h"
+#include "tensorflow/core/framework/variant_op_registry.h"
+#include "tensorflow/core/kernels/cwise_ops_common.h"
+#include "tensorflow/core/kernels/dense_update_functor.h"
+#include "tensorflow/core/kernels/fill_functor.h"
+#include "tensorflow/core/kernels/mkl/mkl_matmul_ops_common.h"
+#include "tensorflow/core/kernels/sparse/kernels.h"
+#include "tensorflow/core/kernels/sparse/mat_mul_op.h"
+#include "tensorflow/core/kernels/sparse/sparse_matrix.h"
+#include "tensorflow/core/kernels/sparse/transpose_op.h"
+#include "tensorflow/core/kernels/transpose_functor.h"
+#include "tensorflow/core/lib/gtl/inlined_vector.h"
+#include "tensorflow/core/platform/threadpool.h"
+#include "tensorflow/core/util/mkl_util.h"
+#include "unsupported/Eigen/CXX11/Tensor"
+
+using dnnl::stream;
+
+namespace tensorflow {
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+// Implements a kernel which, given a SparseMatrix `a` and dense Tensor `b`,
+// computes a dense Tensor `c` satisfying `c = a * b` where * denotes matrix
+// multiplication.
+//
+// The rank of both `a` and `b` must be equal and their shapes must be
+// compatible for matrix multiplication. Otherwise, InvalidArgument runtime
+// errors will be thrown. Only inputs of rank 2 are supported.
+//
+template <typename Device, typename T, bool native_format = false>
+class MklSparseMatrixMatMulOp : public MklDnnMatMulOpBase<T, void, T> {
+ private:
+  tensorflow::CSRMatMulCPUOp<T> eigen_sparse_matmul_op_;
+
+ public:
+  explicit MklSparseMatrixMatMulOp(OpKernelConstruction* ctx)
+      : MklDnnMatMulOpBase<T, void, T>(ctx), eigen_sparse_matmul_op_(ctx) {}
+
+  // Throws errors if there are issues with the input.
+  Status ValidateInputs(const CSRSparseMatrix& sparse_matrix_a,
+                        const Tensor& dense_tensor_b, int* rank) {
+    // Validate datatypes.
+    if (sparse_matrix_a.dtype() != dense_tensor_b.dtype()) {
+      return errors::InvalidArgument(
```

Comment:
Use absl::InvalidArgument (along with absl::StrCat to build up the error message).  Here and throughout.

Reply:
Sure thing, just fixed. Changed all files that this PR touches.